### DISCUSSION TITLE
Treat Prisma schema the same as Terraform state

### DISCUSSION THREAD

**Discussion by danburonline** [prisma/prisma#15322] [Q&A] [answered]
I've been working a lot lately with Terraform, an Infrastructure as Code (IaC) tool from HashiCorp that makes it easy to describe cloud infrastructures in declarative code. Now when I came back to Prisma, I realised that the Prisma schema file is similar to a Terraform file: You describe the structure and content declaratively, and a tool (in this case, Prisma) converts it into the actual implementation; a database.

![image](https://user-images.githubusercontent.com/34251194/190195615-2dce91d1-0baa-4f1e-9333-f981d7831ee8.png)

While thinking about this, I wondered why Prisma doesn't treat its schema file the same way Terraform treats its lock file, with the option to store it on what they call a "state backend". This state backend allows the Terraform state to be stored in, say, an S3 bucket on AWS, making it accessible to multiple environments (e.g. prod, staging, etc.). Wouldn't such a handling of the schema files in Prisma be great? Because for example, I would like to reuse the same schema (or parts of it) in different microservices accessing the same database instead of using one big schema in a monolith or defining smaller schemas per microservice in the [database-per-microservice](https://microservices.io/patterns/data/database-per-service.html) pattern.

I think such a treatment would make using Prisma in microservice environments much more attractive, as there are other clients/generators for other languages. I am curious what you think. ✌️ 

**Context:** There is an [open ticket](https://github.com/prisma/prisma/issues/1563) asking for removed schemas. Unfortunately, this ticket has not been updated for quite some time. Also: There is [a blog post](https://isidoro-ferreiro.medium.com/share-your-prisma-client-across-projects-44d1c7aca6fd) describing how to create an NPM package with the schema stored in it as a workaround that the company I work for will use.

**Comment by nikolasburk** [comment, marked as answer]
Hey Daniel, thanks so much for sharing this idea with us, definitely very interesting! I've forwarded it internally to our Product team and will nudge them to give you some feedback on this 🙌

**Reply by danburonline** [reply]
Amazing!
